### PR TITLE
Clear string pointers before returning them to the memory pool.

### PR DESCRIPTION
**Pull request by notbdu**
String pointers are now cleared before putting string arrays back into the memory pool.

### CODE REVIEW

File: x/pool/ref_counted_pooled_string_array_test.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+	"github.com/xichen2020/eventdb/x/strings"
```
Comment (xichen2020) [inline x/pool/ref_counted_pooled_string_array_test.go]: Import

File: index/field/docs_field.go [hunk 2]
```diff
@@ -871,11 +872,12 @@ func (b *docsFieldBuilder) addTime(docID int32, v int64) error {
...
-	intArrayPool    *pool.BucketizedIntArrayPool
-	doubleArrayPool *pool.BucketizedFloat64ArrayPool
-	stringArrayPool *pool.BucketizedStringArrayPool
-	int64ArrayPool  *pool.BucketizedInt64ArrayPool
+	boolArrayPool       *pool.BucketizedBoolArrayPool
+	intArrayPool        *pool.BucketizedIntArrayPool
+	doubleArrayPool     *pool.BucketizedFloat64ArrayPool
+	stringArrayPool     *pool.BucketizedStringArrayPool
+	int64ArrayPool      *pool.BucketizedInt64ArrayPool
+	stringValuesResetFn strings.ClearArrayFn
```
Comment (xichen2020) [inline index/field/docs_field.go]: Perhaps just `ArrayFn`?

File: index/field/docs_field.go [hunk 3]
```diff
@@ -963,3 +965,15 @@ func (o *DocsFieldBuilderOptions) SetInt64ArrayPool(v *pool.BucketizedInt64Array
 func (o *DocsFieldBuilderOptions) Int64ArrayPool() *pool.BucketizedInt64ArrayPool {
 	return o.int64ArrayPool
 }
+
+// SetStringValuesResetFn sets a value reset function for string values.
+func (o *DocsFieldBuilderOptions) SetStringValuesResetFn(fn strings.ClearArrayFn) *DocsFieldBuilderOptions {
```
Comment (xichen2020) [inline index/field/docs_field.go]: nit: `SetStringArrayResetFn`.

File: index/field/docs_field.go [hunk 4]
```diff
@@ -963,3 +965,15 @@ func (o *DocsFieldBuilderOptions) SetInt64ArrayPool(v *pool.BucketizedInt64Array
...
+	opts := *o
+	opts.stringValuesResetFn = fn
+	return &opts
+}
+
+// StringValuesResetFn resets string values before returning a string array back to the memory pool.
+func (o *DocsFieldBuilderOptions) StringValuesResetFn() strings.ClearArrayFn {
```
Comment (xichen2020) [inline index/field/docs_field.go]: `StringArrayResetFn`.

File: storage/mutable_segment.go [hunk 5]
```diff
@@ -100,7 +101,8 @@ func newMutableSegment(
 		SetIntArrayPool(opts.IntArrayPool()).
 		SetDoubleArrayPool(opts.DoubleArrayPool()).
 		SetStringArrayPool(opts.StringArrayPool()).
-		SetInt64ArrayPool(opts.Int64ArrayPool())
+		SetInt64ArrayPool(opts.Int64ArrayPool()).
+		SetStringValuesResetFn(strings.ValuesResetFn)
```
Comment (xichen2020) [inline storage/mutable_segment.go]: `ResetArrayFn`.

Comment (xichen2020) [reply storage/mutable_segment.go]: Also add a comment here to say this is to avoid holding on to documents received after the array is returned to pool.

File: x/pool/ref_counted_pooled_string_array_test.go [hunk 6]
```diff
@@ -15,7 +18,7 @@ func TestRefCountedPooledStringArray(t *testing.T) {
 	pool.Init(func(capacity int) []string { return make([]string, 0, capacity) })
 
 	vals := pool.Get(4)
-	arr := NewRefCountedPooledStringArray(vals, pool)
+	arr := NewRefCountedPooledStringArray(vals, pool, strings.ValuesResetFn)
```
Comment (xichen2020) [inline x/pool/ref_counted_pooled_string_array_test.go]: Perhaps just leave as nil so you don't need to change the other parts of test?

File: x/strings/strings.go [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package strings
+
+// ClearArrayFn is the type signature of a string slice clearing fn.
+type ClearArrayFn func(values []string)
```
Comment (xichen2020) [inline x/strings/strings.go]: Just `ArrayFn`? More generic that way.

File: x/strings/strings.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+// ValuesResetFn clears string pointers inside of a slice of strings.
+func ValuesResetFn(values []string) {
```
Comment (xichen2020) [inline x/strings/strings.go]: Just `ResetArray`?

File: x/pool/ref_counted_pooled_string_array_test.go [hunk 9]
```diff
@@ -32,24 +33,25 @@ func TestRefCountedPooledStringArray(t *testing.T) {
 	require.Equal(t, 4, len(arr.Get()))
 	require.Equal(t, int32(2), arr2.cnt.RefCount())
 	require.Equal(t, 4, len(arr2.Get()))
-	expected1 := []string{"foo", "bar", "baz", "cat"}
-	assertReturnedToStringArrayPool(t, pool, expected1, false)
+	assertReturnedToStringArrayPool(t, pool, arr.Get(), false)
```
Comment (xichen2020) [inline x/pool/ref_counted_pooled_string_array_test.go]: Hm I don't think you need to change the rest of the tests since you already set the reset fn to nil above. I generally prefer to assert against a very explicit value if possible so in this case I prefer to assert against an expected string array instead.
